src/protocol/check_zero.rs

```diff
@@ -0,0 +1,149 @@
+/// First, the parties generate a secret sharing of a random value `r`, which is not known to any of them.
+/// Next, the parties compute a secret sharing of `r * v` using a multiplication protocol that is secure up to an additive attack.
+/// Then, the parties reveal (sometimes called "open") the secret sharing of `r * v`
+/// If `v` was a secret sharing of zero, this revealed value will also be zero.
+/// On the other hand, if `v` was NOT a secret sharing of zero, then there are two possibilities:
+/// 1.) If the randomly chosen value `r` just so happenned to be zero, then the revealed value will be zero.
+/// This will happen with probability `1/|F|` (where `|F|` denotes the cardinality of the field)
+/// 2.) If the randomly chosen value `r` is any other value in the field aside from zero, then the revealed value will NOT be zero.
+///
+/// Clearly, the accuracy of this protocol is highly dependent on the field that is used.
+/// In a large field, like the integers modulo 2^31 - 1, the odds of `check_zero` returning "true"
+/// when `v` is NOT actually a sharing of zero is extrmely small; it is less than one two billion odds.
+/// In a silly field, like our test field of the integers modulo 31, the odds are very good. It'll incorrectly return "true"
+/// about 3.2% of the time.
+///
+/// ## Errors
+/// Lots of things may go wrong here, from timeouts to bad output. They will be signalled
+/// back via the error response
+#[allow(dead_code)]
+pub async fn check_zero<F: Field, N: Network>(
+    ctx: ProtocolContext<'_, N>,
+    record_id: RecordId,
+    v: Replicated<F>,
+) -> Result<bool, BoxError> {
+    let prss = &ctx.prss();
+    let (left, right) = prss.generate_fields(record_id.into());
+    let r_sharing = Replicated::new(left, right);
+
+    let r_times_v = ctx
+        .narrow(&Step::MultiplyWithR)
+        .multiply(record_id)
+        .await
+        .execute(r_sharing, v)
+        .await?;
+    let r_revealed = reveal(ctx.narrow(&Step::RevealR), record_id, r_times_v).await?;
+
+    Ok(r_revealed == F::ZERO)
+}
+
+#[cfg(test)]
+pub mod tests {
+    use crate::error::BoxError;
+    use crate::field::{Field, Fp31};
+    use crate::protocol::{check_zero::check_zero, QueryId, RecordId};
+    use crate::test_fixture::{logging, make_contexts, make_world, share, TestWorld};
+
+    #[tokio::test]
+    async fn basic() -> Result<(), BoxError> {
+        logging::setup();
+
+        let world: TestWorld = make_world(QueryId);
+        let context = make_contexts(&world);
+        let mut rng = rand::thread_rng();
+
+        for i in 0..u32::from(Fp31::PRIME) {
+            let a = Fp31::from(i);
+            let mut num_false_positives = 0;
+            for j in 0..10 {
+                let a_shares = share(a, &mut rng);
+                let row = i * u32::from(Fp31::PRIME) + j;
+                let record_id = RecordId::from(row);
```

Comment:
Can you instead just maintain a counter for this? i.e.,

```rust
let mut record_id = 0;
for i in ... {
    for j in ... {
        record_id += 1;
        let record_id = RecordId::from(record_id); // or name it something else because this shadowing is super sketchy...
    }
}
```

Reply:
yup, I'll just use a counter of some sort.